raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java

```diff
@@ -272,20 +280,103 @@ public void testUpdateHighWatermarkQuorumSizeThree() {
         assertTrue(state.updateReplicaState(node1, 0, new LogOffsetMetadata(20L)));
         assertEquals(Optional.of(new LogOffsetMetadata(20L)), state.highWatermark());
         assertFalse(state.updateReplicaState(node2, 0, new LogOffsetMetadata(20L)));
         assertEquals(Optional.of(new LogOffsetMetadata(20L)), state.highWatermark());
     }
 
+    @Test
+    public void testUpdateHighWatermarkAddingFollowerToVoterStates() {
+        int node1 = 1;
+        int node2 = 2;
+        Set<Integer> originalVoterSet = mkSet(localId, node1);
+        LeaderState<?> state = newLeaderState(originalVoterSet, 10L);
+        assertFalse(state.updateLocalState(new LogOffsetMetadata(15L), originalVoterSet));
+        assertFalse(state.updateReplicaState(node1, 0, new LogOffsetMetadata(10L)));
+        assertEquals(Optional.empty(), state.highWatermark());
+
+        // updating replica state of 2 before it joins voterSet should not increase HW to 15L
+        assertFalse(state.updateReplicaState(node2, 0, new LogOffsetMetadata(15L)));
+        assertEquals(Optional.empty(), state.highWatermark());
+
+        // adding node2 to voterSet will cause HW to increase to 15L
+        Set<Integer> voterSetWithNode2 = mkSet(localId, node1, node2);
+        assertTrue(state.updateLocalState(new LogOffsetMetadata(15L), voterSetWithNode2));
+        assertEquals(Optional.of(new LogOffsetMetadata(15L)), state.highWatermark());
+
+        // HW will not update to 16L until a majority reaches it
+        assertFalse(state.updateLocalState(new LogOffsetMetadata(16L), voterSetWithNode2));
+        assertEquals(Optional.of(new LogOffsetMetadata(15L)), state.highWatermark());
+        assertTrue(state.updateReplicaState(node2, 0, new LogOffsetMetadata(16L)));
+        assertEquals(Optional.of(new LogOffsetMetadata(16L)), state.highWatermark());
+    }
+
+    @Test
+    public void testUpdateHighWatermarkRemovingFollowerFromVoterStates() {
+        int node1 = 1;
+        int node2 = 2;
+        Set<Integer> originalVoterSet = mkSet(localId, node1, node2);
+        LeaderState<?> state = newLeaderState(originalVoterSet, 10L);
+        assertFalse(state.updateLocalState(new LogOffsetMetadata(15L), originalVoterSet));
+        assertTrue(state.updateReplicaState(node1, 0, new LogOffsetMetadata(15L)));
+        assertFalse(state.updateReplicaState(node2, 0, new LogOffsetMetadata(10L)));
+        assertEquals(Optional.of(new LogOffsetMetadata(15L)), state.highWatermark());
+
+        // removing node1 should not decrement HW to 10L
+        Set<Integer> voterSetWithoutNode1 = mkSet(localId, node2);
+        assertFalse(state.updateLocalState(new LogOffsetMetadata(17L), voterSetWithoutNode1));
+        assertEquals(Optional.of(new LogOffsetMetadata(15L)), state.highWatermark());
+
+        // HW cannot change until after node2 catches up to last HW
+        assertFalse(state.updateReplicaState(node2, 0, new LogOffsetMetadata(14L)));
```

Comment:
I think we can also increase leader's LEO to 16 and show that it doesn't increase the HWM.

Reply:
leader's LEO at this point is 17 (line 330L), however I can bump it up one further to demonstrate the HW doesn't change